File: Smodr/src/main/java/com/kevintcoughlin/smodr/services/MediaService.java
package com.kevintcoughlin.smodr.services;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.IBinder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.kevintcoughlin.smodr.models.Item;

public final class MediaService extends Service implements MediaPlayer.OnErrorListener,
        MediaPlayer.OnPreparedListener {
    private static final int THIRTY_SECONDS_IN_MILLISECONDS = 30000;
    @NonNull
    public static final String INTENT_EPISODE_URL = "intent_episode_url";
    @NonNull
    public static final String INTENT_EPISODE_TITLE = "intent_episode_title";
    @NonNull
    public static final String INTENT_EPISODE_DESCRIPTION = "intent_episode_description";
    @NonNull
    public static final String ACTION_PLAY = "com.kevintcoughlin.smodr.app.PLAY";
    @NonNull
    public static final String ACTION_PAUSE = "com.kevintcoughlin.smodr.app.PAUSE";
    @NonNull
    public static final String ACTION_RESUME = "com.kevintcoughlin.smodr.app.RESUME";
    @NonNull
    public static final String ACTION_STOP = "com.kevintcoughlin.smodr.app.STOP";
    @NonNull
    public static final String ACTION_FORWARD = "com.kevintcoughlin.smodr.app.FORWARD";
    @NonNull
    public static final String ACTION_REWIND = "com.kevintcoughlin.smodr.app.REWIND";
    @Nullable
    private MediaPlayer mMediaPlayer;

    public static Intent createIntent(@NonNull Context context, @NonNull final Item item) {
        final Intent intent = new Intent(context, MediaService.class);
        final String mediaUrlString = item.getUri().toString();

        intent.setAction(MediaService.ACTION_PLAY);
        intent.putExtra(MediaService.INTENT_EPISODE_URL, mediaUrlString);
        intent.putExtra(MediaService.INTENT_EPISODE_TITLE, item.title);
        intent.putExtra(MediaService.INTENT_EPISODE_DESCRIPTION, item.description);

        return intent;
    }

    public static Intent createAction(@NonNull Context context, @NonNull final String action) {
        final Intent intent = new Intent(context, MediaService.class);

        intent.setAction(action);

        return intent;
    }

    public int onStartCommand(final Intent intent, final int flags, final int startId) {
        super.onStartCommand(intent, flags, startId);
        // This could be null currently
        final String url = intent.getStringExtra(INTENT_EPISODE_URL);
        final String action = intent.getAction();

        if (action != null) {
            switch (action) {
                case ACTION_PAUSE:
                    pausePlayback();
                    break;
                case ACTION_PLAY:
                    startPlayback(url);
                    break;
                case ACTION_RESUME:
                    resumePlayback();
                    break;
                case ACTION_STOP:
                    stopPlayback();
                    break;
                case ACTION_FORWARD:
                    seekTo(THIRTY_SECONDS_IN_MILLISECONDS);
                    break;
                case ACTION_REWIND:
                    seekTo(-THIRTY_SECONDS_IN_MILLISECONDS);
                    break;
            }
        }

        return Service.START_REDELIVER_INTENT;
    }

    private void resumePlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.start();
        }
    }

    private void startPlayback(final String url) {
        if (mMediaPlayer != null && mMediaPlayer.isPlaying()) {
            stopPlayback();
        }

        final Uri uri = Uri.parse(url);

        mMediaPlayer = MediaPlayer.create(this, uri);
        mMediaPlayer.start();
    }

    private void seekTo(final int milliseconds) {
        if (mMediaPlayer != null) {
            mMediaPlayer.seekTo(mMediaPlayer.getCurrentPosition() + milliseconds);
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        stopPlayback();
        return true;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopPlayback();
        if (mMediaPlayer != null) {
            mMediaPlayer.release();
        }
    }

    private void pausePlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.pause();
        }
//        stopForeground(true);
    }

    private void stopPlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.stop();
        }
//        stopForeground(true);
    }


    @Override
    public void onPrepared(final MediaPlayer mediaPlayer) {
        resumePlayback();
    }
}

File: Smodr/src/main/java/com/kevintcoughlin/smodr/services/MediaService.kt
package com.kevintcoughlin.smodr.services

import android.app.Service
import android.content.Intent
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.media.MediaPlayer.OnPreparedListener
import android.net.Uri
import android.os.Binder
import android.os.IBinder
import android.util.Log
import com.kevintcoughlin.smodr.services.MediaService.IPlaybackListener
import org.jetbrains.annotations.Contract
import kotlin.math.min

internal interface IMediaService {
    fun seekTo(milliseconds: Int)
    val isPlaying: Boolean
    fun resumePlayback()
    fun pausePlayback()
    fun stopPlayback()
    fun forward()
    fun rewind()
    fun setPlaybackListener(listener: IPlaybackListener?)
}

class MediaService : Service(), MediaPlayer.OnErrorListener, OnPreparedListener, IMediaService, OnCompletionListener {
    override fun onCompletion(mp: MediaPlayer) {
        mListener?.onCompletion()
    }

    interface IPlaybackListener {
        fun onStartPlayback()
        fun onStopPlayback()
        fun onCompletion()
    }

    private var mMediaPlayer: MediaPlayer? = null
    private val mBinder: IBinder = MediaServiceBinder()
    private var mListener: IPlaybackListener? = null

    inner class MediaServiceBinder : Binder() {
        val service: MediaService
            get() = this@MediaService
    }

    @Contract(pure = true)
    override fun onBind(intent: Intent): IBinder {
        return mBinder
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        val url = Uri.parse(intent.getStringExtra(INTENT_EPISODE_URL))
        val action = intent.action
        if (action != null) when (action) {
            ACTION_PAUSE -> pausePlayback()
            ACTION_PLAY -> startPlayback(url)
            ACTION_RESUME -> resumePlayback()
            ACTION_STOP -> stopPlayback()
            ACTION_FORWARD -> rewind()
            ACTION_REWIND -> forward()
        }
        return START_REDELIVER_INTENT
    }

    override fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
        stopPlayback()
        return true
    }

    override fun onDestroy() {
        super.onDestroy()
        stopPlayback()
        mMediaPlayer?.release()
        mListener = null
    }

    override fun seekTo(milliseconds: Int) {
        mMediaPlayer?.seekTo(milliseconds)
    }

    override val isPlaying: Boolean
        get() = mMediaPlayer?.isPlaying == true

    val duration: Int
        get() = mMediaPlayer?.duration ?: -1

    val currentTime: Int
        get() = mMediaPlayer?.currentPosition ?: -1

    val remainingTime: Int
        get() = duration - currentTime

    override fun resumePlayback() {
        mMediaPlayer?.start()
        mListener?.onStartPlayback()
    }

    override fun pausePlayback() {
        mMediaPlayer?.pause()
        mListener?.onStopPlayback()
    }

    override fun stopPlayback() {
        mMediaPlayer?.stop()
        mListener?.onStopPlayback()
    }

    override fun forward() {
        val position = mMediaPlayer?.currentPosition?.plus(THIRTY_SECONDS_IN_MILLISECONDS)
        position?.let { mMediaPlayer?.duration?.let { it1 -> min(it, it1) } }?.let { seekTo(it) }
    }

    override fun rewind() {
        mMediaPlayer?.currentPosition?.minus(THIRTY_SECONDS_IN_MILLISECONDS)?.let { seekTo(it) }
    }

    override fun setPlaybackListener(listener: IPlaybackListener?) {
        mListener = listener
    }

    override fun onPrepared(mediaPlayer: MediaPlayer) {
        resumePlayback()
    }

    fun startPlayback(url: Uri?) {
        if (mMediaPlayer?.isPlaying == true) {
            stopPlayback()
        }
        try {
            mMediaPlayer = MediaPlayer.create(this, url)
            mMediaPlayer?.setOnCompletionListener(this)
            mMediaPlayer?.start()
            mListener?.onStartPlayback()
        } catch (exception: NullPointerException) {
            Log.e("MediaService", exception.message, exception)
        }
    }

    companion object {
        const val THIRTY_SECONDS_IN_MILLISECONDS = 30000
        const val INTENT_EPISODE_URL = "intent_episode_url"
        const val ACTION_PLAY = "com.com.kevintcoughlin.smodr.app.PLAY"
        const val ACTION_PAUSE = "com.com.kevintcoughlin.smodr.app.PAUSE"
        const val ACTION_RESUME = "com.com.kevintcoughlin.smodr.app.RESUME"
        const val ACTION_STOP = "com.com.kevintcoughlin.smodr.app.STOP"
        const val ACTION_FORWARD = "com.com.kevintcoughlin.smodr.app.FORWARD"
        const val ACTION_REWIND = "com.com.kevintcoughlin.smodr.app.REWIND"
    }
}